syntax/rust.vim

```diff
@@ -153,6 +153,14 @@ syn region    rustDerive      start="derive(" end=")" contained contains=rustDer
 " Some are deprecated (Encodable, Decodable) or to be removed after a new snapshot (Show).
 syn keyword   rustDeriveTrait contained Clone Hash RustcEncodable RustcDecodable Encodable Decodable PartialEq Eq PartialOrd Ord Rand Show Debug Default FromPrimitive Send Sync Copy
 
+" dyn keyword: It's only a keyword when used inside a type expression,
+" with two options: detected by a preceding '<' or, or a preceding '&'
+" as long the next token is an identifier. This is so that uses of dyn 
+" variable names such as in 'let &dyn = &2' and 'let dyn = 2' will not 
+" get highlighted as a keyword.
+syn match     rustKeyword "[&<] *\(dyn\) \+[a-zA-Z_]"me=e-1,he=e-1 contains=rustOperator,rustDynKeyword,rustSigil
```

Comment:
A type can have multiple generic parameters, e.g. `Foo<dyn Bar, dyn Baz>`. Adding `,` to `[&<]` would resolve that.

But then there are other things that can precede it as well, e.g. `mut` (`&mut dyn Foo`) or a lifetime (`&'foo dyn Foo`).

In normal syntax it’s an error to have a path following an identifier, so I think just dropping the preceding token part would be sufficient. I think we can ignore paths with leading double colon (`dyn ::std::fmt::Display`), since that’s on the way out in Rust 2018 anyway. Weirdly-placed comments are always going to mess up a simple highlighter like this, so let’s not try to go crazy about that.

I’d prefer to use the start of rustIdentifier rather than `[A-Za-z_]`; that way we keep support for non-ASCII identifiers.

All up, I suggest `syn match rustKeyword "\<dyn\ze\_s\+\%([^[:cntrl:][:space:][:punct:][:digit:]]\|_\)" contains=rustDynKeyword` (and a tweaking of the comment to match, of course).